{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "Bioinformatics.ipynb",
      "provenance": [],
      "collapsed_sections": [],
      "mount_file_id": "1spa6WIJ7AnqFEx23npocnCrRULcEQRwb",
      "authorship_tag": "ABX9TyNks1rzj0xG3nkTdtJxmCsw",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "accelerator": "TPU"
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/Baek-Donghyeon/Bioinformatics-Algorithms/blob/main/Bioinformatics.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Z0vB1kWjeVkM"
      },
      "source": [
        "# Bioinformatics"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "HKZGqBuzeXD3",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "cd4c8cdb-3d3c-4fb2-ce0e-b119fdc96eb0"
      },
      "source": [
        "import numpy\n",
        "import random\n",
        "import re\n",
        "import sys\n",
        "from collections import Counter\n",
        "from collections import defaultdict\n",
        "import itertools\n",
        "from functools import reduce\n",
        "import copy\n",
        "sys.setrecursionlimit(3000)\n",
        "\n",
        "\n",
        "def k_mer_of(text, k):\n",
        "  for i in range(len(text) - k + 1):\n",
        "    yield text[i:i+k]\n",
        "\n",
        "\n",
        "def hamming_distance(pattern, text):\n",
        "  # Minimum distance between pattern and all k_mers in text\n",
        "  k = len(pattern)\n",
        "  distance = k\n",
        "  for k_mer in k_mer_of(text, k):\n",
        "    distance = min(distance, sum(pattern[j] != k_mer[j] for j in range(k)))\n",
        "  return distance\n",
        "\n",
        "\n",
        "def pattern_to_number(pattern):\n",
        "  symbol_to_number = {'A': 0, 'C': 1, 'G': 2, 'T': 3}\n",
        "  if len(Pattern) == 0:\n",
        "      return 0\n",
        "  return 4*pattern_to_number(pattern[:-1]) + symbol_to_number[pattern[-1]]\n",
        "\n",
        "\n",
        "def number_to_pattern(number, k):\n",
        "  number_to_symbol = {0: 'A', 1: 'C', 2: 'G', 3: 'T'}\n",
        "  try:\n",
        "    if k == 1:\n",
        "      return number_to_symbol(number)\n",
        "    return number_to_pattern(number // 4, k-1) + number_to_symbol[number % 4]\n",
        "  except:\n",
        "    raise ValueError()\n",
        "\n",
        "\n",
        "def count_pattern(text, pattern):\n",
        "  return sum(k_mer == pattern for k_mer in k_mer_of(text, len(pattern)))\n",
        "\n",
        "\n",
        "def find_frequent_k_mer(text, k, distance = 0, reverse_complement = False):\n",
        "  ''' k   length of the k_mer '''\n",
        "  frequency_dict = dict(frequency_dict(text, k, distance, reverse_complement))\n",
        "  max = max(frequency_dict, key = lambda x : frequency_dict[x])\n",
        "  return list(filter(lambda x : x[1] == max, frequency_dict.items())) \n",
        "\n",
        "\n",
        "def frequency_dict(text, k, distance = 0, reverse_complement = False):\n",
        "  ''' k   length of the k_mer '''\n",
        "  if reverse_complement:\n",
        "    return frequency_dict(reverse_complement(text), k, distance) + frequency_dict(text, k, distance)\n",
        "  else:\n",
        "    return Counter([neighbor for k_mer in k_mer_of(text, k) for neighbor in neighbor(k_mer, distance)])\n",
        "\n",
        "\n",
        "def reverse_complement(text):\n",
        "  compliment_dict = {'A': 'T', 'T': 'A', 'G': 'C', 'C': 'G'}\n",
        "  return ''.join(reversed([compliment_dict[base] for base in text]))\n",
        "\n",
        "\n",
        "def find_pattern_occurence(text, pattern, distance = 0):\n",
        "  # hamming_distance(text, pattern) â‰¤ d\n",
        "  k = len(pattern)\n",
        "  return [i for i in range(len(text) - k + 1) if hamming_distance(text[i:i+k], pattern) <= distance]\n",
        "\n",
        "\n",
        "def find_clump(text, k, L, t):\n",
        "  '''\n",
        "  k   length of the k_mer\n",
        "  L   length of an interval\n",
        "  t   min ocurrences of a pattern in the interval\n",
        "  '''\n",
        "  position_dict = defaultdict(list)\n",
        "  frequency_dict = frequency_dict(text, k)\n",
        "  for pattern, frequency in frequency_dict.items():\n",
        "    if frequency >= t:\n",
        "      position_dict[pattern] = find_occurrences_of_pattern(text, pattern)\n",
        "  for pattern, position in position_dict.items():\n",
        "    for i in range(len(position) - t + 1):\n",
        "      if position[i+t-1] - position[i] <= L - k + 1:\n",
        "        yield pattern\n",
        "\n",
        "\n",
        "def find_min_skew_position(text):\n",
        "  # skew : abs(#'G'-#'C')\n",
        "  skew = [0]\n",
        "  for base in text:\n",
        "    if base == 'C': skew.append(skew[-1]-1)\n",
        "    elif base == 'G': skew.append(skew[-1]+1)\n",
        "    else: skew.append(skew[-1])\n",
        "  min = min(skew)\n",
        "  return list(filter(lambda x : x == min, skew)) # return minimum skew\n",
        "\n",
        "\n",
        "def neighbor(pattern, distance):\n",
        "  length = len(pattern)\n",
        "  mismatch_base = itertools.product(\"ACTG\", repeat = distance) # all possible 'distance' length letter strings\n",
        "  for positions in itertools.combinations(iter(range(length)), distance):\n",
        "    for base in mismatch_base:\n",
        "      neighbor = copy.deepcopy(pattern)\n",
        "      for i in distance:\n",
        "        neighbor[position[i]] = base[i]\n",
        "        yield neighbor\n",
        "      \n",
        "\n",
        "def enumerate_motif(strings, k, distance):\n",
        "  # Find all (k, distance)-motifs in a collection of strings.\n",
        "  neighbor_list = []\n",
        "  for text in strings:\n",
        "    neighbor_list.append(set(neighbor(text, k, distance)))\n",
        "  motifs = neighbor_list[0]\n",
        "  for neighbor in neighbor_list:\n",
        "    motifs = motifs & neighbor\n",
        "  return list(motifs)\n",
        "\n",
        "\n",
        "def find_median_string(strings, k, distance = 0):\n",
        "  # A median string for Dna minimizes d(Pattern, Dna) over all k-mers Pattern.\n",
        "  motifs = enumerate_motif(strings, k, distance)\n",
        "  if not motifs:\n",
        "    find_median_string(strings, k, distance + 1)\n",
        "  else:\n",
        "    motifs_dict = defaultdict(int)\n",
        "    for pattern in motifs:\n",
        "      motifs_dict[pattern] = sum(hamming_distance(pattern, text) for text in strings)\n",
        "    return min(motifs_dict, key = lambda x : motifs_dict[x]) # return key that minimizes the value\n",
        "\n",
        "\n",
        "def find_profile_probable_k_mer(text, k, profile):\n",
        "  # A k-mer that was most likely to have been generated by Profile among all k-mers in Text.\n",
        "  score = 0\n",
        "  for k_mer in k_mer_of(text, k):\n",
        "    score = max(score, score(profile, [k_mer]))\n",
        "  return score\n",
        "\n",
        "\n",
        "def score(profile, motif):\n",
        "  return sum(reduce(lambda x, y: x*y, list(profile[pattern_to_number(k_mer[i])][i]\n",
        "              for i in range(k))) for k_mer in motif)\n",
        "\n",
        "\n",
        "def form_profile(motif):\n",
        "  k = len(motif[0])\n",
        "  profile = numpy.full((4,k), 1) # pseudocount\n",
        "  for i in range(k):\n",
        "    for k_mer in motif:\n",
        "      Profile[pattern_to_number(k_mer[i])][i] += 1\n",
        "  return profile/(len(motif)+4)\n",
        "\n",
        "\n",
        "def greedy_motif_search(strings, k):\n",
        "  best_motif = [text[:k] for text in strings]\n",
        "  for i in range(len(strings[0]) - k + 1):\n",
        "    motif = [strings[0][i:i+k]]\n",
        "    for j in range(1,len(strings)):\n",
        "      profile = form_profile(motif)\n",
        "      motif.append(find_profile_probable_k_mer(strings[j], k, motif))\n",
        "    if score(profile, motif) < score(profile, best_motif):\n",
        "      best_motif = motif\n",
        "  return best_motif\n",
        "\n",
        "\n",
        "# def RandomizedMotifSearch(Dna, k, t):\n",
        "#   w = len(Dna[0])\n",
        "#   # from Symbol to Number representation\n",
        "#   NumDna = [[SymbolToNumber(Symbol) for Symbol in list(line)] for line in Dna]\n",
        "#   Motifs = []\n",
        "#   for i in range(t):\n",
        "#     j = random.randrange(w-k+1)\n",
        "#     Motifs.append(NumDna[i][j:j+k])\n",
        "#   BestMotifs = Motifs\n",
        "#   for x in range(1000):\n",
        "#     while True:\n",
        "#       Profile = ProfileFormation(Motifs, k)\n",
        "#       NewMotifs = []\n",
        "#       for Text in NumDna:\n",
        "#         NewMotifs.append(MostProbableKmer(Text, k, Profile))\n",
        "#       Motifs = NewMotifs\n",
        "#       if Score(Motifs, Profile) > Score(BestMotifs, Profile):\n",
        "#         BestMotifs = Motifs\n",
        "#       else:\n",
        "#         break\n",
        "#     Motifs = []\n",
        "#     for i in range(t):\n",
        "#       j = random.randrange(w-k+1)\n",
        "#       Motifs.append(NumDna[i][j:j+k])\n",
        "#   # from Number to Symbol representation\n",
        "#   return [''.join([NumberToSymbol(Number) for Number in line]) for line in BestMotifs]\n",
        "\n",
        "\n",
        "# def GibbsSampler(Dna, k, t, N):\n",
        "#   w = len(Dna[0])\n",
        "#   # from Symbol to Number representation\n",
        "#   NumDna = [[SymbolToNumber(Symbol) for Symbol in list(line)] for line in Dna]\n",
        "#   Motifs = []\n",
        "#   for i in range(t):\n",
        "#     j = random.randrange(w-k+1)\n",
        "#     Motifs.append(NumDna[i][j:j+k])\n",
        "#   BestMotifs = Motifs\n",
        "#   for j in range(N):\n",
        "#     i = random.randrange(t)\n",
        "#     Motifs.pop(i)\n",
        "#     Profile = ProfileFormation(Motifs, k)\n",
        "#     #Profile-randomly generated k-mer in a string Text.\n",
        "#     Motifs.insert(i, MostProbableKmer(NumDna[i], k, Profile))\n",
        "#     if Score(Motifs, Profile) > Score(BestMotifs, Profile):\n",
        "#       BestMotifs = Motifs\n",
        "#   return [''.join([NumberToSymbol(Number) for Number in line]) for line in BestMotifs]\n",
        "\n",
        "\n",
        "def find_string(strings, d = None):\n",
        "  first = strings[0]\n",
        "  if d != None: # Paired-Reads\n",
        "    half = len(first)//2\n",
        "    return find_string([front[:half] for front in strings]) + find_string([rear[half:] for rear in strings[-half-d:]])\n",
        "  return first[:-1] + ''.join([text[-1] for text in strings])\n",
        "\n",
        "\n",
        "def de_brujin_graph(strings):\n",
        "  graph = defaultdict(list)\n",
        "  if len(strings[0]) == 2: # Paired-Reads\n",
        "    for f,r in strings: # ex) GAGA|TTGA, f = GAGA, r = TTGA\n",
        "      graph[f[:-1]+r[:-1]].append(f[1:]+r[1:])         \n",
        "  else:\n",
        "    for k_mer in strings:\n",
        "      graph[k_mer[:-1]].append(k_mer[1:])\n",
        "  return graph\n",
        "\n",
        "\n",
        "def eulerian(graph):\n",
        "\n",
        "  # determine whether it is an eulerian path or cycle\n",
        "  graph = copy.deepcopy(graph)\n",
        "  in_degree = Counter(list(itertools.chain(*graph.values())))\n",
        "  out_degree = Counter([k for k,v in graph.items() for __ in range(len(v))])\n",
        "  odd_degree = (out_degree - in_degree).keys()\n",
        "  if odd_degree:\n",
        "    stack = list(odd_degree) # start from odd_degree / eulerian path\n",
        "  else:\n",
        "    stack = [list(graph.keys())[0]] # start from arbitrary node / eulerian cycle\n",
        "\n",
        "  # DFS\n",
        "  visit = []\n",
        "  while stack:\n",
        "    try:\n",
        "      stack.append(graph[stack[-1]].pop())\n",
        "    except: # backtrack\n",
        "      visit.append(stack.pop())\n",
        "  visit.reverse()\n",
        "  return visit\n",
        "\n",
        "\n",
        "def reconstruct_string(strings, d = None):\n",
        "  # d = # of bases b/w front and rear ends\n",
        "  return find_string(eulerian(de_brujin_graph(strings)),d)\n",
        "\n",
        "\n",
        "def find_k_universal_circular_string(k):\n",
        "  # all possible k_mers\n",
        "  # [:-k+1] : for circular_string\n",
        "  return reconstruct_string(list(map(''.join, product('01', repeat=k))))[:-k+1]\n",
        "\n",
        "\n",
        "def generate_contig(strings):\n",
        "  graph = de_brujin_graph(strings)\n",
        "  in_degree = Counter(list(itertools.chain(*graph.values())))\n",
        "  out_degree = Counter([k for k,v in graph.items() for __ in range(len(v))])\n",
        "  collection = []\n",
        "  branch = set(node for node in graph.keys() if in_degree[node] != 1 or out_degree[node] != 1)\n",
        "  branch.update(set(node for values in graph.values() for node in values if in_degree[node] != 1 or out_degree[node] != 1))\n",
        "  for node in branch:\n",
        "    while graph[node]:\n",
        "      contig = [node, graph[node].pop()]\n",
        "      while contig[-1] not in branch:\n",
        "        contig.append(graph[contig[-1]].pop())\n",
        "      collection.append(find_string(contig))\n",
        "  return collection\n",
        "\n",
        "\n",
        "f = open('/content/drive/My Drive/Colab Notebooks/input.txt', 'r')\n",
        "strings = []\n",
        "for line in f.readlines():\n",
        "  strings.append(line.strip())\n",
        "for contig in sorted(generate_contig(strings)):\n",
        "  print(contig)\n",
        "f.close()\n",
        "\n",
        "  # def ContigGeneration(PathGraph):\n",
        "  # #Generate the contigs from a collection of reads (with imperfect coverage).\n",
        "  # #Non-branching if in(v) = out(v) = 1 for each intermediate node v of this path,\n",
        "  #   Graph = []\n",
        "  #   AdjDict = {}\n",
        "  #   DegreeDict = {} # Degree : # of nodes [In, Out]\n",
        "  #   for Nodes in PathGraph:\n",
        "  #     for Node in Nodes:\n",
        "  #       AdjDict.setdefault(Node)\n",
        "  #       DegreeDict.setdefault(Node, [0,0])\n",
        "  #   for Nodes in PathGraph:\n",
        "  #     AdjDict[Nodes[0]] = Nodes[1:]\n",
        "  #     DegreeDict[Nodes[0]][1] = len(Nodes[1:])\n",
        "\n",
        "  #   for EndNodes in AdjDict.values():\n",
        "  #     if EndNodes:\n",
        "  #       for EndNode in EndNodes:\n",
        "  #         DegreeDict[EndNode][0] = (DegreeDict[EndNode][0]+1)\n",
        "\n",
        "  #   def Search(key, contig):\n",
        "  #     value = AdjDict[key].pop()\n",
        "  #     contig.append(value)\n",
        "  #     if DegreeDict[value] == [1,1] and AdjDict[value]:\n",
        "  #       return Search(value, contig)\n",
        "  #     else: # Finish Search if value is a branch point or has no out\n",
        "  #       return Reconstruct(contig)\n",
        "    \n",
        "  #   Contigs = []\n",
        "  #   IsNotEmpty = True\n",
        "  #   while IsNotEmpty:\n",
        "  #     IsNotEmpty = False\n",
        "  #     for key in AdjDict.keys():\n",
        "  #       if AdjDict[key] and not DegreeDict[key] == [1,1]:\n",
        "  #       # key is a branch point\n",
        "  #         Contigs.append(Search(key, [key]))\n",
        "  #         IsNotEmpty = True\n",
        "\n",
        "  #   return Contigs\n",
        "\n",
        "\n",
        "  # def ProteinTranslation(Pattern):\n",
        "  #   Protein = []\n",
        "  #   for i in range(len(Pattern)//3):\n",
        "  #     Codon = Pattern[3*i:3*i+3]\n",
        "  #     if Codon in ['UAA','UAG','UGA']:\n",
        "  #       break\n",
        "  #     Protein.append(GeneticCode[Codon])\n",
        "  #   return ''.join(Protein)\n",
        "\n",
        "\n",
        "  # def DnaToRna(Dna):\n",
        "  #   return re.sub('T','U',Dna)\n",
        "\n",
        "\n",
        "  # def PeptideEncoding(Text, Peptide):\n",
        "  # #Find substrings of a genome encoding a given amino acid sequence.\n",
        "  #   Substrings = []\n",
        "  #   PeptideLen = len(Peptide)\n",
        "  #   for i in range(len(Text)-3*PeptideLen+1):\n",
        "  #     substring = Text[i:i+3*PeptideLen]\n",
        "  #     if ProteinTranslation(DnaToRna(substring)) == Peptide or ProteinTranslation(DnaToRna(ReverseCompliment(substring))) == Peptide:\n",
        "  #       Substrings.append(substring)\n",
        "  #   return Substrings\n",
        "\n",
        "\n",
        "  # def Cyclospectrum(Peptide):\n",
        "  # #Generate the theoretical spectrum of a cyclic peptide.\n",
        "  # #The theoretical spectrum of a cyclic peptide Peptide is the collection of all of the masses of its subpeptides\n",
        "  #   PeptideLen = len(Peptide)\n",
        "  #   TheoreticalSpectrum = [0,Mass(Peptide)]\n",
        "\n",
        "  #   Peptide += Peptide\n",
        "  #   for i in range(1,PeptideLen):\n",
        "  #     for j in range(PeptideLen):\n",
        "  #       TheoreticalSpectrum.append(Mass(Peptide[j:j+i]))\n",
        "    \n",
        "  #   return sorted(TheoreticalSpectrum)\n",
        "\n",
        "\n",
        "  # def Mass(Peptide):\n",
        "  #   Mass = 0\n",
        "  #   for AminoAcid in Peptide:\n",
        "  #     Mass += IntegerMass[AminoAcid]\n",
        "  #   return Mass\n",
        "\n",
        "\n",
        "  # def BFCyclopeptideSequencing(Mass):\n",
        "  # #Compute the number of peptides of given total mass.\n",
        "  #   if Mass in ReducedMass.values():\n",
        "  #     global cnt\n",
        "  #     cnt += 1\n",
        "  #   elif Mass > 57:\n",
        "  #     for AminoAcid in ReducedMass.values():\n",
        "  #       if Mass > AminoAcid:\n",
        "  #         BFCyclopeptideSequencing(Mass-AminoAcid)\n",
        "\n",
        "\n",
        "  # def CyclopeptideSequencing(Spectrum):\n",
        "    \n",
        "  #   def Expand(CandidatePeptides):\n",
        "  #     Expand = []\n",
        "  #     for Peptide in CandidatePeptides:\n",
        "  #       for AminoAcid in ReducedMass.keys():\n",
        "  #         Expand.append(Peptide+AminoAcid)\n",
        "  #     return Expand\n",
        "  #   CandidatePeptides = list(ReducedMass.keys())\n",
        "  #   FinalPeptides = []\n",
        "  #   while True:\n",
        "  #     CopyCandidatePeptides = CandidatePeptides.copy()\n",
        "  #     for Peptide in CopyCandidatePeptides:\n",
        "  #       Peptidespectrum = Cyclospectrum(Peptide)\n",
        "  #       if Peptidespectrum == Spectrum:\n",
        "  #         FinalPeptides.append(Peptide)\n",
        "  #         CandidatePeptides.remove(Peptide)\n",
        "  #       else:\n",
        "  #         for Fragment in Peptidespectrum:\n",
        "  #           if Fragment not in Spectrum:\n",
        "  #             CandidatePeptides.remove(Peptide)\n",
        "  #             break\n",
        "  #     if not CandidatePeptides:\n",
        "  #       break\n",
        "  #     print(CandidatePeptides)\n",
        "  #     CandidatePeptides = Expand(CandidatePeptides)\n",
        "  #   print(FinalPeptides)\n",
        "  #   return FinalPeptides\n",
        "\n",
        "\n",
        "  # GeneticCode = {'AAA':'K','AAC':'N','AAG':'K','AAU':'N','ACA':'T','ACC':'T','ACG':'T','ACU':'T',\n",
        "  #               'AGA':'R','AGC':'S','AGG':'R','AGU':'S','AUA':'I','AUC':'I','AUG':'M','AUU':'I',\n",
        "  #               'CAA':'Q','CAC':'H','CAG':'Q','CAU':'H','CCA':'P','CCC':'P','CCG':'P','CCU':'P',\n",
        "  #               'CGA':'R','CGC':'R','CGG':'R','CGU':'R','CUA':'L','CUC':'L','CUG':'L','CUU':'L',\n",
        "  #               'GAA':'E','GAC':'D','GAG':'E','GAU':'D','GCA':'A','GCC':'A','GCG':'A','GCU':'A',\n",
        "  #               'GGA':'G','GGC':'G','GGG':'G','GGU':'G','GUA':'V','GUC':'V','GUG':'V','GUU':'V',\n",
        "  #               'UAA':'*','UAC':'Y','UAG':'*','UAU':'Y','UCA':'S','UCC':'S','UCG':'S','UCU':'S',\n",
        "  #               'UGA':'*','UGC':'C','UGG':'W','UGU':'C','UUA':'L','UUC':'F','UUG':'L','UUU':'F'}\n",
        "\n",
        "  # IntegerMass = {'G':57,'A':71,'S':87,'P':97,'V':99,'T':101,'C':103,'I':113,'L':113,'N':114,\n",
        "  #               'D':115,'K':128,'Q':128,'E':129,'M':131,'H':137,'F':147,'R':156,'Y':163,'W':186}\n",
        "\n",
        "  # ReducedMass = {'G':57,'A':71,'S':87,'P':97,'V':99,'T':101,'C':103,'I':113,'N':114,\n",
        "  #               'D':115,'K':128,'E':129,'M':131,'H':137,'F':147,'R':156,'Y':163,'W':186}\n",
        "  #Spectrum = list(map(int,f.readline().split(' ')))\n",
        "  # print(CyclopeptideSequencing(Spectrum))\n",
        "\n",
        "\n"
      ],
      "execution_count": 154,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "AAAAACTTGAAGGCTTAT\n",
            "AAAAACTTGAAGGCTTAT\n",
            "AAAACTTGAAGGCTTATG\n",
            "AAAACTTGAAGGCTTATG\n",
            "AAAAGTGGGCGCGCGCCG\n",
            "AAAAGTGGGCGCGCGCCG\n",
            "AAACTTGAAGGCTTATGGCCCCAGTGCGTA\n",
            "AAAGTGGGCGCGCGCCGC\n",
            "AAAGTGGGCGCGCGCCGC\n",
            "AACATTAATAGCTGCTCTAAGCCCTCCGCA\n",
            "AACGAAAACCTAAAGGCT\n",
            "AACGAAAACCTAAAGGCT\n",
            "AACGGCCCCGGGCGCCCTCATCGACCTGTTA\n",
            "AACTCTCCACCCGGGCTG\n",
            "AACTCTCCACCCGGGCTG\n",
            "AAGAAAACGTTGATATGTCTCTGAGCATGAT\n",
            "AAGAATTGGCCAGCGAGA\n",
            "AAGAATTGGCCAGCGAGA\n",
            "AAGCGAAATCTTCTCTAA\n",
            "AAGCGAAATCTTCTCTAA\n",
            "AAGCGTCAGCGAGATGCGGAGGCCAACGACAGG\n",
            "AAGTGGGCGCGCGCCGCAAGAAAACGTTGATA\n",
            "AAGTGGTTGAAAACAAGTCGAAGTGTTATGGATAG\n",
            "AATACACCCTGCAGAGTATTACTTTAGTCCCTTA\n",
            "AATCGTTCAGGCTACGAA\n",
            "AATCGTTCAGGCTACGAA\n",
            "AATCTGCGATGGATAGAGGGGATGCCCTTCGGG\n",
            "ACCAGGTACGGATATCTG\n",
            "ACCAGGTACGGATATCTG\n",
            "ACCCCGCTAGTGGACCTG\n",
            "ACCCCGCTAGTGGACCTG\n",
            "ACCCTGGTGCAGAAGAGG\n",
            "ACCCTGGTGCAGAAGAGG\n",
            "ACCGTAGCGCGCGCGGCA\n",
            "ACCGTAGCGCGCGCGGCA\n",
            "ACGAAAACCTAAAGGCTGAGTGTCTACTTGGGTT\n",
            "ACGCGGAGCGTCTGTGGT\n",
            "ACGCGGAGCGTCTGTGGT\n",
            "ACGTGTTTCATACCCCAT\n",
            "ACGTGTTTCATACCCCAT\n",
            "ACGTTTCCCCGAGTCTGAAGTGGTTGAAAACAA\n",
            "ACTATGTGTCGGACCGGA\n",
            "ACTATGTGTCGGACCGGA\n",
            "ACTCTCCACCCGGGCTGTAACATTAATAGCT\n",
            "ACTTTTCCGTGCTTTTCG\n",
            "ACTTTTCCGTGCTTTTCG\n",
            "AGAAAAACTTGAAGGCTT\n",
            "AGAAAAACTTGAAGGCTT\n",
            "AGAAAGCTTCATCGGAGCAGTCGGCCTAATGTT\n",
            "AGAATTGGCCAGCGAGACTTTTCCGTGCTTTT\n",
            "AGCACTTGTGCGCGTCCT\n",
            "AGCACTTGTGCGCGTCCT\n",
            "AGCGAAATCTTCTCTAAG\n",
            "AGCGAAATCTTCTCTAAG\n",
            "AGCGAGCCAAGATCGTCGTTCTCAGGCTACAAAGT\n",
            "AGGCCAACGACAGGATTCTGTAGCATTCAAGC\n",
            "AGGGGATGCCCTTCGGGA\n",
            "AGGGGATGCCCTTCGGGA\n",
            "AGGTTCCGAACACCACGA\n",
            "AGGTTCCGAACACCACGA\n",
            "AGTATTTGCTCATCTTAT\n",
            "AGTATTTGCTCATCTTAT\n",
            "AGTCGAGAGGGTGGCGACCAGGTACGGATATC\n",
            "AGTCGGCCTAATGTTTCG\n",
            "AGTCGGCCTAATGTTTCG\n",
            "AGTGTCTACTTGGGTTACGTGACCATGTAACGGCG\n",
            "ATACCGTAGCGCGCGCGG\n",
            "ATACCGTAGCGCGCGCGG\n",
            "ATACGGAAGACCCTCCCT\n",
            "ATACGGAAGACCCTCCCT\n",
            "ATAGCACTTGTGCGCGTC\n",
            "ATAGCACTTGTGCGCGTC\n",
            "ATCCTAGAGTCAGGGCGT\n",
            "ATCCTAGAGTCAGGGCGT\n",
            "ATCGGCGTATGTTATACT\n",
            "ATCGGCGTATGTTATACT\n",
            "ATCGTTCAGGCTACGAAA\n",
            "ATCGTTCAGGCTACGAAA\n",
            "ATGGAGCCCACAAACATC\n",
            "ATGGAGCCCACAAACATC\n",
            "ATGGCCCCAGTGCGTAAA\n",
            "ATGGCCCCAGTGCGTAAA\n",
            "ATGTAAGAGATACCACTTACGTAGCTCGGAAA\n",
            "ATGTCTCTGAGCATGATG\n",
            "ATGTCTCTGAGCATGATG\n",
            "ATTACTTTAGTCCCTTAA\n",
            "ATTACTTTAGTCCCTTAA\n",
            "ATTCATTAGGTCGGGAGC\n",
            "ATTCATTAGGTCGGGAGC\n",
            "ATTCTGGGGTTCAGCAGG\n",
            "ATTCTGGGGTTCAGCAGG\n",
            "CAAAAGTGGGCGCGCGCC\n",
            "CAAAAGTGGGCGCGCGCC\n",
            "CAAGAAAACGTTGATATG\n",
            "CAAGAAAACGTTGATATG\n",
            "CACGCGGAGCGTCTGTGG\n",
            "CACGCGGAGCGTCTGTGG\n",
            "CACTTGTGCGCGTCCTTGTAGCAAACTTTG\n",
            "CAGGTACGGATATCTGTGTTGTTATGTTGACA\n",
            "CAGTATTTGCTCATCTTA\n",
            "CAGTATTTGCTCATCTTA\n",
            "CAGTCGGCCTAATGTTTC\n",
            "CAGTCGGCCTAATGTTTC\n",
            "CATATACTCCTCATGCGAATACACCCTGCAGA\n",
            "CATGTAAGAGATACCACT\n",
            "CATGTAAGAGATACCACT\n",
            "CATGTATATTCACTTTCCATATACTCCTCAT\n",
            "CCAGGTACGGATATCTGT\n",
            "CCAGGTACGGATATCTGT\n",
            "CCATATACTCCTCATGCG\n",
            "CCATATACTCCTCATGCG\n",
            "CCCCGCTAGTGGACCTGCGGAGGACGGGAAAA\n",
            "CCCCTGTACGTGCCGACG\n",
            "CCCCTGTACGTGCCGACG\n",
            "CCCCTGTTTGATTGTGCCTGCAGCTCCCGCG\n",
            "CCCGAACGGCCCCGGGCG\n",
            "CCCGAACGGCCCCGGGCG\n",
            "CCCTCATCGACCTGTTAC\n",
            "CCCTCATCGACCTGTTAC\n",
            "CCCTCTCGCCTTCACGTG\n",
            "CCCTCTCGCCTTCACGTG\n",
            "CCCTGGTGCAGAAGAGGCGCCCCTGTTTGATTGT\n",
            "CCCTGTACGTGCCGACGG\n",
            "CCCTGTACGTGCCGACGG\n",
            "CCGAACGGCCCCGGGCGC\n",
            "CCGAACGGCCCCGGGCGC\n",
            "CCGCCCACTTCGTCTATC\n",
            "CCGCCCACTTCGTCTATC\n",
            "CCGGACCCGTAATTGATTCTGGGGTTCAGC\n",
            "CCGTAGCGCGCGCGGCAGTATTTGCTCATCTT\n",
            "CCTAGAGTCAGGGCGTGCAAGTTCAAGTGCAG\n",
            "CCTCATCGACCTGTTACT\n",
            "CCTCATCGACCTGTTACT\n",
            "CCTCTCGCCTTCACGTGCCGCCCACTTCGTCTA\n",
            "CCTGCAGCTCCCGCGTGG\n",
            "CCTGCAGCTCCCGCGTGG\n",
            "CCTGTACGTGCCGACGGT\n",
            "CCTGTACGTGCCGACGGT\n",
            "CCTTGTAGCAAACTTTGT\n",
            "CCTTGTAGCAAACTTTGT\n",
            "CGAAACGTGTAGTGGTCT\n",
            "CGAAACGTGTAGTGGTCT\n",
            "CGAACGGCCCCGGGCGCC\n",
            "CGAACGGCCCCGGGCGCC\n",
            "CGAAGTGTTATGGATAGT\n",
            "CGAAGTGTTATGGATAGT\n",
            "CGAATACACCCTGCAGAG\n",
            "CGAATACACCCTGCAGAG\n",
            "CGAGCGAGGGTAGCAGAAATCGTTCAGGCTACGA\n",
            "CGATACGGAAGACCCTCC\n",
            "CGATACGGAAGACCCTCC\n",
            "CGCCCACTTCGTCTATCG\n",
            "CGCCCACTTCGTCTATCG\n",
            "CGCCCCTGTTTGATTGTG\n",
            "CGCCCCTGTTTGATTGTG\n",
            "CGCGGAGCGTCTGTGGTA\n",
            "CGCGGAGCGTCTGTGGTA\n",
            "CGCTGGCCGAGCCATATC\n",
            "CGCTGGCCGAGCCATATC\n",
            "CGGAAGTATAGGGTTATT\n",
            "CGGAAGTATAGGGTTATT\n",
            "CGGAGGACGGGAAAACAC\n",
            "CGGAGGACGGGAAAACAC\n",
            "CGGAGGCCAACGACAGGA\n",
            "CGGAGGCCAACGACAGGA\n",
            "CGGTCGCCCAGGTTTGGC\n",
            "CGGTCGCCCAGGTTTGGC\n",
            "CGTAACCGCCCAAAGTACGCTGGCCGAGCCATAT\n",
            "CGTGTTTCATACCCCATGAGTCGAGAGGGTGGC\n",
            "CGTTGGTGGTCTACAAGGACCCTGGTGCAGAAGAG\n",
            "CTAAGCCCTCCGCAGAAAAACTTGAAGGCT\n",
            "CTATGTGTCGGACCGGAATTCATTAGGTCGGGAG\n",
            "CTCAGGCTACAAAGTTCTGTGCATGACGGCAGAG\n",
            "CTCATCGACCTGTTACTTACGTTTCCCCGAGT\n",
            "CTCTAAGCCCTCCGCAGA\n",
            "CTCTAAGCCCTCCGCAGA\n",
            "CTCTGAGCATGATGGGAATCTGCGATGGATAG\n",
            "CTCTGGCTCACGGCAACT\n",
            "CTCTGGCTCACGGCAACT\n",
            "CTGCAGCTCCCGCGTGGC\n",
            "CTGCAGCTCCCGCGTGGC\n",
            "CTGGCCGAGCCATATCCCCTGTACGTGCCGA\n",
            "CTGGCTCACGGCAACTTA\n",
            "CTGGCTCACGGCAACTTA\n",
            "CTGTAACATTAATAGCTG\n",
            "CTGTAACATTAATAGCTG\n",
            "CTGTACGTGCCGACGGTCCCCAGAGCATGGCC\n",
            "CTGTAGCATTCAAGCCCGAACGGCCCCGGGC\n",
            "CTTACGTAGCTCGGAAAG\n",
            "CTTACGTAGCTCGGAAAG\n",
            "CTTACGTTTCCCCGAGTC\n",
            "CTTACGTTTCCCCGAGTC\n",
            "CTTGTAGCAAACTTTGTT\n",
            "CTTGTAGCAAACTTTGTT\n",
            "CTTTTCCGTGCTTTTCGA\n",
            "CTTTTCCGTGCTTTTCGA\n",
            "GAAAAACTTGAAGGCTTA\n",
            "GAAAAACTTGAAGGCTTA\n",
            "GAAACGTGTAGTGGTCTTCGTTGGTGGTCTACAA\n",
            "GAACGGCCCCGGGCGCCC\n",
            "GAACGGCCCCGGGCGCCC\n",
            "GAAGTATAGGGTTATTTTTTGGACTAACTCAA\n",
            "GAAGTGGTTGAAAACAAG\n",
            "GAAGTGGTTGAAAACAAG\n",
            "GAAGTGTTATGGATAGTGGAGAAAGCTTCATCGGA\n",
            "GAATACACCCTGCAGAGT\n",
            "GAATACACCCTGCAGAGT\n",
            "GAATCTGCGATGGATAGA\n",
            "GAATCTGCGATGGATAGA\n",
            "GACCAGGTACGGATATCT\n",
            "GACCAGGTACGGATATCT\n",
            "GACGTGTTTCATACCCCA\n",
            "GACGTGTTTCATACCCCA\n",
            "GACTTTTCCGTGCTTTTC\n",
            "GACTTTTCCGTGCTTTTC\n",
            "GAGAAAGCTTCATCGGAG\n",
            "GAGAAAGCTTCATCGGAG\n",
            "GAGGACGGGAAAACACCCTCTCGCCTTCACGT\n",
            "GAGGCCAACGACAGGATT\n",
            "GAGGCCAACGACAGGATT\n",
            "GAGTCGAGAGGGTGGCGA\n",
            "GAGTCGAGAGGGTGGCGA\n",
            "GAGTGTCTACTTGGGTTA\n",
            "GAGTGTCTACTTGGGTTA\n",
            "GATACGGAAGACCCTCCC\n",
            "GATACGGAAGACCCTCCC\n",
            "GATCCTAGAGTCAGGGCG\n",
            "GATCCTAGAGTCAGGGCG\n",
            "GATTCTGGGGTTCAGCAG\n",
            "GATTCTGGGGTTCAGCAG\n",
            "GCAAGAAAACGTTGATAT\n",
            "GCAAGAAAACGTTGATAT\n",
            "GCAAGTTCAAGTGCAGCACGCGGAGCGTCTGTG\n",
            "GCACTTGTGCGCGTCCTT\n",
            "GCACTTGTGCGCGTCCTT\n",
            "GCAGTCGGCCTAATGTTT\n",
            "GCAGTCGGCCTAATGTTT\n",
            "GCATAGTTGCACTCAGCCGCGAGCGAGGGTAGCAG\n",
            "GCATGTATATTCACTTTC\n",
            "GCATGTATATTCACTTTC\n",
            "GCCCACTTCGTCTATCGGAAGTATAGGGTTAT\n",
            "GCCCCTGTTTGATTGTGC\n",
            "GCCCCTGTTTGATTGTGC\n",
            "GCCGCCCACTTCGTCTAT\n",
            "GCCGCCCACTTCGTCTAT\n",
            "GCCGGACCCGTAATTGAT\n",
            "GCCGGACCCGTAATTGAT\n",
            "GCCTGCAGCTCCCGCGTG\n",
            "GCCTGCAGCTCCCGCGTG\n",
            "GCGAAACGTGTAGTGGTC\n",
            "GCGAAACGTGTAGTGGTC\n",
            "GCGAAATCTTCTCTAAGCGAAACGTGTAGTGGT\n",
            "GCGAGCGAGGGTAGCAGA\n",
            "GCGAGCGAGGGTAGCAGA\n",
            "GCGGAGCGTCTGTGGTAACTCTCCACCCGGGC\n",
            "GCGGAGGACGGGAAAACA\n",
            "GCGGAGGACGGGAAAACA\n",
            "GCTCTAAGCCCTCCGCAG\n",
            "GCTCTAAGCCCTCCGCAG\n",
            "GCTGGCCGAGCCATATCC\n",
            "GCTGGCCGAGCCATATCC\n",
            "GGAAGTATAGGGTTATTT\n",
            "GGAAGTATAGGGTTATTT\n",
            "GGAGCCCACAAACATCGTGGCGGTACAGTTGTT\n",
            "GGAGGACGGGAAAACACC\n",
            "GGAGGACGGGAAAACACC\n",
            "GGAGGCCAACGACAGGAT\n",
            "GGAGGCCAACGACAGGAT\n",
            "GGATGCCCTTCGGGACGAACCATTAGCAGCGCGC\n",
            "GGCATAGTTGCACTCAGC\n",
            "GGCATAGTTGCACTCAGC\n",
            "GGCCCCAGTGCGTAAATAAGCGTCAGCGAGATG\n",
            "GGCCGGACCCGTAATTGA\n",
            "GGCCGGACCCGTAATTGA\n",
            "GGCGGTACAGTTGTTGATCCTAGAGTCAGGG\n",
            "GGGATGCCCTTCGGGACG\n",
            "GGGATGCCCTTCGGGACG\n",
            "GGGGATGCCCTTCGGGAC\n",
            "GGGGATGCCCTTCGGGAC\n",
            "GGTCGCCCAGGTTTGGCA\n",
            "GGTCGCCCAGGTTTGGCA\n",
            "GGTTCCGAACACCACGAC\n",
            "GGTTCCGAACACCACGAC\n",
            "GTAACATTAATAGCTGCT\n",
            "GTAACATTAATAGCTGCT\n",
            "GTATATGTGTTTATGGTC\n",
            "GTATATGTGTTTATGGTC\n",
            "GTATTTGCTCATCTTATACTATGTGTCGGACCGG\n",
            "GTCCCCAGAGCATGGCCC\n",
            "GTCCCCAGAGCATGGCCC\n",
            "GTCGCCCAGGTTTGGCAAAAGTGGGCGCGCGC\n",
            "GTCGGCCTAATGTTTCGGTCGCCCAGGTTTGG\n",
            "GTCTCTGAGCATGATGGG\n",
            "GTCTCTGAGCATGATGGG\n",
            "GTGACCATGTAACGGCGA\n",
            "GTGACCATGTAACGGCGA\n",
            "GTGCATGACGGCAGAGGTAACGAAAACCTAAAGG\n",
            "GTGCATGTATATTCACTT\n",
            "GTGCATGTATATTCACTT\n",
            "GTGGCCGGACCCGTAATT\n",
            "GTGGCCGGACCCGTAATT\n",
            "GTGGCGGTACAGTTGTTG\n",
            "GTGGCGGTACAGTTGTTG\n",
            "GTGTTGTTATGTTGACAG\n",
            "GTGTTGTTATGTTGACAG\n",
            "GTTCATGTAAGAGATACC\n",
            "GTTCATGTAAGAGATACC\n",
            "GTTCCGAACACCACGACA\n",
            "GTTCCGAACACCACGACA\n",
            "GTTGTTATGTTGACAGGGGCATAGTTGCACTCAG\n",
            "TAACATTAATAGCTGCTC\n",
            "TAACATTAATAGCTGCTC\n",
            "TAACGAAAACCTAAAGGC\n",
            "TAACGAAAACCTAAAGGC\n",
            "TAACTCTCCACCCGGGCT\n",
            "TAACTCTCCACCCGGGCT\n",
            "TAAGCGTCAGCGAGATGC\n",
            "TAAGCGTCAGCGAGATGC\n",
            "TACCGTAGCGCGCGCGGC\n",
            "TACCGTAGCGCGCGCGGC\n",
            "TACGGAAGACCCTCCCTCGTAACCGCCCAAAGT\n",
            "TACGTTTCCCCGAGTCTG\n",
            "TACGTTTCCCCGAGTCTG\n",
            "TAGCACTTGTGCGCGTCC\n",
            "TAGCACTTGTGCGCGTCC\n",
            "TATATGTGTTTATGGTCGATACGGAAGACCCTC\n",
            "TATGGCCCCAGTGCGTAA\n",
            "TATGGCCCCAGTGCGTAA\n",
            "TCATGTAAGAGATACCAC\n",
            "TCATGTAAGAGATACCAC\n",
            "TCATTAGGTCGGGAGCAAGCGAAATCTTCTCTA\n",
            "TCCATATACTCCTCATGC\n",
            "TCCATATACTCCTCATGC\n",
            "TCCCCAGAGCATGGCCCTCTGGCTCACGGCAAC\n",
            "TCCCCTGTACGTGCCGAC\n",
            "TCCCCTGTACGTGCCGAC\n",
            "TCCTAGAGTCAGGGCGTG\n",
            "TCCTAGAGTCAGGGCGTG\n",
            "TCGGCGTATGTTATACTGACCCCGCTAGTGGACCT\n",
            "TCGTAACCGCCCAAAGTA\n",
            "TCGTAACCGCCCAAAGTA\n",
            "TCGTTCAGGCTACGAAATGGAGCCCACAAACAT\n",
            "TCGTTGGTGGTCTACAAG\n",
            "TCGTTGGTGGTCTACAAG\n",
            "TCTAAGCCCTCCGCAGAA\n",
            "TCTAAGCCCTCCGCAGAA\n",
            "TCTCAGGCTACAAAGTTC\n",
            "TCTCAGGCTACAAAGTTC\n",
            "TCTCTGAGCATGATGGGA\n",
            "TCTCTGAGCATGATGGGA\n",
            "TCTGGCTCACGGCAACTT\n",
            "TCTGGCTCACGGCAACTT\n",
            "TCTGTAGCATTCAAGCCC\n",
            "TCTGTAGCATTCAAGCCC\n",
            "TGACCATGTAACGGCGACGTATATGTGTTTATGGT\n",
            "TGAGTCGAGAGGGTGGCG\n",
            "TGAGTCGAGAGGGTGGCG\n",
            "TGATCCTAGAGTCAGGGC\n",
            "TGATCCTAGAGTCAGGGC\n",
            "TGATTCTGGGGTTCAGCA\n",
            "TGATTCTGGGGTTCAGCA\n",
            "TGCAAGTTCAAGTGCAGC\n",
            "TGCAAGTTCAAGTGCAGC\n",
            "TGCAGCTCCCGCGTGGCCGGACCCGTAAT\n",
            "TGCATGTATATTCACTTT\n",
            "TGCATGTATATTCACTTT\n",
            "TGCCTGCAGCTCCCGCGT\n",
            "TGCCTGCAGCTCCCGCGT\n",
            "TGCGGAGGACGGGAAAAC\n",
            "TGCGGAGGACGGGAAAAC\n",
            "TGGAGCCCACAAACATCG\n",
            "TGGAGCCCACAAACATCG\n",
            "TGGCCCCAGTGCGTAAAT\n",
            "TGGCCCCAGTGCGTAAAT\n",
            "TGGCCGGACCCGTAATTG\n",
            "TGGCCGGACCCGTAATTG\n",
            "TGGCGGTACAGTTGTTGA\n",
            "TGGCGGTACAGTTGTTGA\n",
            "TGGCTCACGGCAACTTATCGGCGTATGTTATAC\n",
            "TGTAACATTAATAGCTGC\n",
            "TGTAACATTAATAGCTGC\n",
            "TGTAGCAAACTTTGTTCATGTAAGAGATAC\n",
            "TGTCTCTGAGCATGATGG\n",
            "TGTCTCTGAGCATGATGG\n",
            "TGTGCATGACGGCAGAGG\n",
            "TGTGCATGACGGCAGAGG\n",
            "TGTTGTTATGTTGACAGG\n",
            "TGTTGTTATGTTGACAGG\n",
            "TTACGTAGCTCGGAAAGTGCATGTATATTCACT\n",
            "TTACGTTTCCCCGAGTCT\n",
            "TTACGTTTCCCCGAGTCT\n",
            "TTACTTTAGTCCCTTAATACCGTAGCGCGCGCG\n",
            "TTCATGTAAGAGATACCA\n",
            "TTCATGTAAGAGATACCA\n",
            "TTCATTAGGTCGGGAGCA\n",
            "TTCATTAGGTCGGGAGCA\n",
            "TTCCATATACTCCTCATG\n",
            "TTCCATATACTCCTCATG\n",
            "TTCCGAACACCACGACAAGAATTGGCCAGCGAG\n",
            "TTCTCAGGCTACAAAGTT\n",
            "TTCTCAGGCTACAAAGTT\n",
            "TTCTGGGGTTCAGCAGGTTCCGAACACCACG\n",
            "TTCTGTAGCATTCAAGCC\n",
            "TTCTGTAGCATTCAAGCC\n",
            "TTGGACTAACTCAATAGCACTTGTGCGCGT\n",
            "TTGTAGCAAACTTTGTTC\n",
            "TTGTAGCAAACTTTGTTC\n",
            "TTTCCGTGCTTTTCGACGTGTTTCATACCCC\n",
            "TTTGGACTAACTCAATAG\n",
            "TTTGGACTAACTCAATAG\n",
            "TTTTCCGTGCTTTTCGAC\n",
            "TTTTCCGTGCTTTTCGAC\n",
            "TTTTGGACTAACTCAATA\n",
            "TTTTGGACTAACTCAATA\n",
            "TTTTTGGACTAACTCAAT\n",
            "TTTTTGGACTAACTCAAT\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "0IOLJRo7kpL3",
        "outputId": "d6e91e21-5ff1-433e-d6cd-a91332fc409a"
      },
      "source": [
        "+for front, rear in [[1,2],[2,3]]:\r\n",
        "  print(front, rear)"
      ],
      "execution_count": 73,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "1 2\n",
            "2 3\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "jPTIdW_DcOQN",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "41550c5d-3be2-447f-e7df-e52879ab9e23"
      },
      "source": [
        "from google.colab import drive\n",
        "drive.mount('/content/drive')"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount(\"/content/drive\", force_remount=True).\n"
          ],
          "name": "stdout"
        }
      ]
    }
  ]
}